### PR TITLE
New Cache System

### PR DESCRIPTION
**Pull request by amirsaam**
- New `Cacher` class has been introduced in `Cacher.swift` that uses `DataCache` package
- Removed `ImageCache` class and rewrote the logic and reworked all usages of it
- Exposed `getITunesData` function so any part of PlayCover can use it as a function as it should not be in a class that is unrelated to its functionality
- Online Images now are being cached via CachedAsyncImage package that uses URLCache as memory
- Installed app icons extracted as NSImage and will be updated on each update (fixes old `ImageCache` bug)

### CODE REVIEW

File: PlayCover/Views/App Views/PlayAppView.swift [hunk 1]
```diff
@@ -182,28 +182,30 @@ struct PlayAppConditionalView: View {
...
-                            .shadow(radius: 1)
-                            .padding(.horizontal, 15)
-                            .padding(.vertical, 5)
-                    } placeholder: {
-                        ProgressView()
-                            .progressViewStyle(.circular)
-                            .frame(width: 60, height: 60)
+                    Group {
+                        if let image = appIcon {
+                            Image(nsImage: image)
```
Comment (IsaacMarovitz) [inline PlayCover/Views/App Views/PlayAppView.swift:198]: This takes significantly longer to load than `AsyncImage`/`ImageCache`. When switching to App Library `Image Cache` takes around ~0.5-1 seconds whereas this system takes around ~4-5 seconds.

**Review by IsaacMarovitz — CHANGES_REQUESTED** [review]

File: PlayCover/Views/App Views/PlayAppView.swift (hunk 1 above)
Comment (amirsaam) [reply PlayCover/Views/App Views/PlayAppView.swift:198]: That's on first load I assume, in my tests it was same as ImageCache

**Review by IsaacMarovitz — APPROVED** [review]

File: PlayCover/Utils/Cacher.swift [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+            for icon in assetsExtractor.extractIcons() where checkImageDimensions(icon, bestResImage) {
+                bestResImage = icon
+            }
+        }
+        cache.write(string: compareStr, forKey: compareStr)
+        cache.write(image: bestResImage!, forKey: app.info.bundleIdentifier)
+        return cache.readImage(forKey: app.info.bundleIdentifier)
+    }
+
+    func getLocalIcon(bundleId: String) async -> NSImage? {
```
Comment (KhoraLee) [inline PlayCover/Utils/Cacher.swift]: `resolveLocalIcon()` and `getLocalIcon()` doesn't need to be async function.

File: PlayCover/Views/App Views/StoreAppView.swift [hunk 3]
```diff
@@ -239,8 +258,19 @@ struct StoreAppConditionalView: View {
...
-            iconURL = await ImageCache.getOnlineImageURL(bundleID: app.bundleID,
-                                                         itunesLookup: app.itunesLookup)
+            if !cache.hasData(forKey: app.itunesLookup) {
+                await Cacher().resolveITunesData(app.itunesLookup)
+            }
+            do {
+                itunesResponce = try cache.readCodable(forKey: app.itunesLookup)
+            } catch {
+                print("Read error \(error.localizedDescription)")
+            }
```
Comment (KhoraLee) [inline PlayCover/Views/App Views/StoreAppView.swift]: `itunesResponce = try? cache.readCodable(forKey: app.itunesLookup)`

File: PlayCover/Utils/Cacher.swift [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+    /// `cache.maxCachePeriodInSecond = 7*86400`      // 1 week
+    /// More details: https://github.com/huynguyencong/DataCache/blob/master/README.md
+
+    func resolveITunesData(_ link: String) async {
+        if let refreshedITunesData = await getITunesData(link) {
+            do {
+                try cache.write(codable: refreshedITunesData, forKey: link)
+            } catch {
+                print("Write error \(error.localizedDescription)")
+            }
```
Comment (KhoraLee) [inline PlayCover/Utils/Cacher.swift]: If error is not handled and just ignored then do-catch is excessive 
`try? cache.write(codable: refreshedITunesData, forKey: link)` is enough

File: PlayCover/Views/MenuBarView.swift [hunk 5]
```diff
@@ -93,7 +94,16 @@ struct PlayCoverViewMenuView: Commands {
...
+                DataCache.instance.cleanAll()
+                 URLCache.iconCache.removeAllCachedResponses()
+                 do {
+                     let cacheFolder = FileManager.default.urls(for: .cachesDirectory, in: .userDomainMask).first!
+                      if FileManager.default.fileExists(atPath: cacheFolder.path) {
+                          try FileManager.default.removeItem(at: cacheFolder)
+                      }
+                  } catch {
+                      Log.shared.error(error)
+                  }
```
Comment (KhoraLee) [inline PlayCover/Views/MenuBarView.swift]: Wrong indentation

**Review by KhoraLee — CHANGES_REQUESTED** [review]

File: PlayCover/Views/MenuBarView.swift (hunk 5 above)
Comment (amirsaam) [reply PlayCover/Views/MenuBarView.swift]: Elaborate please. This is the old cache folder for reference

Comment (KhoraLee) [reply PlayCover/Views/MenuBarView.swift]: These lines have one more space than other lines

Comment (amirsaam) [reply PlayCover/Views/MenuBarView.swift]: Oh! I see

**Review by KhoraLee — APPROVED** [review]
LGTM

**Review by Depal1 — APPROVED** [review]

**Pull request merged**